chain/rust/src/builders/tx_builder.rs

```diff
@@ -0,0 +1,4966 @@
+use super::witness_builder::InputAggregateWitnessData;
+use super::redeemer_builder::RedeemerSetBuilder;
+use super::redeemer_builder::RedeemerWitnessKey;
+use super::witness_builder::PlutusScriptWitnessEnum;
+use super::witness_builder::RequiredWitnessSet;
+use super::witness_builder::TransactionWitnessSetBuilder;
+use super::witness_builder::merge_fake_witness;
+use std::collections::{BTreeMap, BTreeSet};
+use rand::Rng;
+
+#[derive(Clone, Default, Debug)]
+struct WitnessBuilders {
+    pub(crate) witness_set_builder: TransactionWitnessSetBuilder,
+    pub(crate) fake_required_witnesses: RequiredWitnessSet,
+    pub(crate) redeemer_set_builder: RedeemerSetBuilder,
+}
+impl WitnessBuilders {
+    fn merge_data(&self, include_fake: bool) -> TransactionWitnessSetBuilder {
+        // add redeemers
+        let redeemers = self.redeemer_set_builder.build(true);
+        let mut witness_set_clone = self.witness_set_builder.clone();
+        witness_set_clone.add_redeemers(&redeemers.unwrap());
+
+        if include_fake {
+            merge_fake_witness(&mut witness_set_clone, &self.fake_required_witnesses);
+            let own_requirements = witness_set_clone.required_wits.clone();
+            merge_fake_witness(&mut witness_set_clone, &own_requirements);
+        }
+
+        witness_set_clone
+    }
+
+    /// build with fake witnesses to estimate tx fee cost
+    pub fn build_fake(&self) -> TransactionWitnessSet {
+        self.merge_data(true).build()
+    }
+
+    /// build without including fake witnesses used for fee estimation. Allows missing witnesses
+    pub fn build_unchecked(&self) -> TransactionWitnessSetBuilder {
+        self.merge_data(false)
+    }
+}
+
+// tx_body must be the result of building from tx_builder
+// constructs the rest of the Transaction using fake witness data of the correct length
+// for use in calculating the size of the final Transaction
+fn fake_full_tx(tx_builder: &TransactionBuilder, body: TransactionBody) -> Result<Transaction, JsError> {
+    Ok(Transaction {
+        body,
+        witness_set: tx_builder.witness_builders.build_fake(),
+        is_valid: true,
+        auxiliary_data: tx_builder.auxiliary_data.clone(),
+    })
+}
+
+fn min_fee(tx_builder: &TransactionBuilder) -> Result<Coin, JsError> {
+    let full_tx = fake_full_tx(tx_builder, tx_builder.build_body()?)?;
+    // we can't know the of scripts yet as they can't be calculated until we build the tx
+    ledger::alonzo::fees::min_no_script_fee(&full_tx, &tx_builder.config.fee_algo)
+}
+
+fn min_fee_with_exunits(tx_builder: &TransactionBuilder) -> Result<Coin, JsError> {
+    let full_tx = fake_full_tx(tx_builder, tx_builder.build_body()?)?;
+    // we can't know the of scripts yet as they can't be calculated until we build the tx
+    ledger::alonzo::fees::min_fee(&full_tx, &tx_builder.config.fee_algo, &tx_builder.config.ex_unit_prices)
+}
+
+#[wasm_bindgen]
+pub enum CoinSelectionStrategyCIP2 {
```

Comment:
The vast majority of the file is for tests. ~1500 is the actual code. It might shrink once rewritten for idiomatic rust code. Here's the approximate breakdown:
* add_change_if_needed: ~300 (free-floating pub function that mutates the tx builder)
* SignedTxBuilder: ~40
* TxRedeemerBuilder: ~80

I don't think per-struct makes sense for breaking it up since the non-tx-builder structs are tiny. You'd have to start breaking off methods off of the tx builder struct itself to make it meaningfully smaller. The coin selection functions are ~250 lines for example.